### PR TITLE
[extractor/medaltv] Fix extraction and account for new url format

### PR DESCRIPTION
**Pull request by xenova**
### Description of your *pull request* and other information

</details>

It would appear as though medal.tv reworked a lot of their front-end code, breaking the current extraction logic for the site. Although a lot of the functionality is the same, we unfortunately have to make an additional request to their /_next API to get the relevant clip data.

I'm not exactly sure when they made these breaking changes, but I was notified of it today (since I was the one who originally added support for the site).

They also changed the URL format:
Old: https://medal.tv/clips/xitf8v3RqmUcd/DlgzpPE0KSmv
New: https://medal.tv/games/fortnite/clips/xitf8v3RqmUcd/DlgzpPE0KSmv

Fortunately, the old links redirect to the new links, so there is backwards compatibility with the current extractor.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/medaltv.py [hunk 1]
```diff
@@ -39,25 +63,46 @@ class MedalTVIE(InfoExtractor):
...
         'only_matching': True,
     }]
 
+    @classmethod
+    def _match_path(cls, url):
+        return cls._match_valid_url(url).group('path')
+
     def _real_extract(self, url):
         video_id = self._match_id(url)
+        path = self._match_path(url)
```
Comment (pukkandan) [inline yt_dlp/extractor/medaltv.py:87]:
```suggestion
    def _real_extract(self, url):
        video_id, path = self._match_valid_url(url).group('id', 'path')
```

File: yt_dlp/extractor/medaltv.py [hunk 2]
```diff
@@ -39,25 +63,46 @@ class MedalTVIE(InfoExtractor):
...
+
         webpage = self._download_webpage(url, video_id)
 
-        hydration_data = self._parse_json(self._search_regex(
-            r'<script[^>]*>\s*(?:var\s*)?hydrationData\s*=\s*({.+?})\s*</script>',
-            webpage, 'hydration data', default='{}'), video_id)
+        next_data = self._parse_json(self._search_regex(
+            r'<script[^>]*__NEXT_DATA__[^>]*>\s*({.+?})\s*</script>',
+            webpage, 'next data', default='{}'), video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/medaltv.py:93]: the recommended way now is to use `_search_json`

File: yt_dlp/extractor/medaltv.py [hunk 3]
```diff
@@ -39,25 +63,46 @@ class MedalTVIE(InfoExtractor):
...
+
+        build_id = next_data.get('buildId')
+        if not build_id:
+            raise ExtractorError(
+                'Could not find build ID.', video_id=video_id)
+
+        locale = next_data.get('locale', 'en')
+
+        api_url = 'https://medal.tv/_next/data/{0}/{1}/{2}/{3}.json'.format(build_id, locale, path, video_id)
+        api_response = self._download_json(api_url, video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/medaltv.py:103]:
```suggestion
        api_response = self._download_json(
            f'https://medal.tv/_next/data/{build_id}/{locale}/{path}/{video_id}.json', video_id)
```

File: yt_dlp/extractor/medaltv.py [hunk 4]
```diff
@@ -39,25 +63,46 @@ class MedalTVIE(InfoExtractor):
...
 
-        clip = try_get(
-            hydration_data, lambda x: x['clips'][video_id], dict) or {}
+        clip = try_get(api_response, lambda x: x['pageProps']['clip'], dict) or {}
```
Comment (pukkandan) [inline yt_dlp/extractor/medaltv.py:105]: This is fine; but fyi, the currently reccomended way is:
```suggestion
        clip = traverse_obj(api_response, ('pageProps', 'clip')) or {}
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/medaltv.py (hunk 3 above)
Comment (xenova) [reply yt_dlp/extractor/medaltv.py:103]: Okay sweet - didn't know if we were allowed to use f-strings (coz of old python support)

**Comment by xenova** [comment]
(Ignore the "Revert" commit, I accidentally clicked "Revert" instead of "View details")

**Pull request merged**